compositor_pipeline/src/queue/queue_thread.rs

```diff
@@ -0,0 +1,258 @@
+                chunks_counter: 0,
+                queue_start_time: start_event.start_time,
+            },
+            video_processor: VideoQueueProcessor {
+                queue: queue_thread.queue,
+                sender: start_event.video_sender,
+                sent_batches_counter: 0,
+                queue_start_time: start_event.start_time,
+            },
+        }
+    }
+
+    fn run(mut self) {
+        let ticker = tick(Duration::from_millis(10));
+
+        loop {
+            ticker.recv().unwrap();
+            loop {
+                let audio_pts_range = self.audio_processor.next_buffer_pts_range();
+                let video_pts = self.video_processor.next_buffer_pts();
+                if video_pts > audio_pts_range.0 {
+                    if self
+                        .audio_processor
+                        .try_push_next_sample_batch(audio_pts_range)
+                        .is_none()
+                    {
+                        break;
+                    }
+                } else if self
+                    .video_processor
+                    .try_push_next_frame(video_pts)
+                    .is_none()
+                {
+                    break;
+                }
+            }
+        }
+    }
+}
+
+struct VideoQueueProcessor {
+    queue: Arc<Queue>,
+    sent_batches_counter: u32,
+    queue_start_time: Instant,
+    sender: Sender<FrameSet<InputId>>,
+}
+
+impl VideoQueueProcessor {
+    fn next_buffer_pts(&self) -> Duration {
+        Duration::from_secs_f64(
+            self.sent_batches_counter as f64 * self.queue.output_framerate.den as f64
+                / self.queue.output_framerate.num as f64,
+        )
+    }
+
+    fn should_push_pts(&self, pts: Duration, queue: &mut MutexGuard<VideoQueue>) -> bool {
+        if !self.queue.ahead_of_time_processing && self.queue_start_time.add(pts) > Instant::now() {
+            return false;
+        }
+        if queue.check_all_inputs_ready_for_pts(pts, self.queue_start_time) {
+            return true;
+        }
+        if queue.has_required_inputs_for_pts(pts, self.queue_start_time) {
+            return false;
+        }
+        self.queue_start_time.add(pts) < Instant::now()
+    }
+
+    fn send_output_frames(&mut self, frames_batch: FrameSet<InputId>, is_required: bool) {
+        if is_required {
+            self.sender.send(frames_batch).unwrap()
+        } else {
+            let send_deadline = self.queue_start_time.add(frames_batch.pts);
+            if self
+                .sender
+                .send_deadline(frames_batch, send_deadline)
+                .is_err()
+            {
+                warn!("Dropping video frame on queue output.");
+            }
+        }
+        self.sent_batches_counter += 1
+    }
+
+    /// Some(()) - Successfully pushed new frame (or dropped it).
+    /// None - Nothing to push.
+    fn try_push_next_frame(&mut self, next_buffer_pts: Duration) -> Option<()> {
```

Comment:
agree on rename, disagree on second point

This might be misleading if try_push_next_frame was public method of the queue, not if it's part of internal implementation accessible only to queue thread.